File: gnm-run-weights/mainTest.cpp
//
// Created by Max Würfel on 26.06.23.
//

#include "GNM.h"
#include <string>
#include <vector>
#include <iostream>
#include "TestData.h"
#include "WeightedGNM.h"


int main() {
    // Load the test data
    std::string pA = "/Users/maxwuerfek/code/diss/gnm-run-weights/testData/A.h5";
    std::string pA_init = "/Users/maxwuerfek/code/diss/gnm-run-weights/testData/Ainit.h5";
    std::string pD = "/Users/maxwuerfek/code/diss/gnm-run-weights/testData/D.h5";

    // Load the synthetic data
    int n = 68, m = 80;
    std::vector<std::vector<double>> A(n, std::vector<double>(n, 0));
    std::vector<std::vector<double>> A_init(n, std::vector<double>(n, 0));
    std::vector<std::vector<double>> D(n, std::vector<double>(n));

    TestData::getSyntheticData(
            A,
            A_init,
            D,
            pA,
            pA_init,
            pD,
            n);



    // Generate parameter space
    double eta = -3.2, gamma = 0.38, alpha = 0.05, omega = 0.9;
    std::vector<std::vector<double>> paramSpace = {{eta, gamma, alpha, omega}};

    // Weighted model parameters
    WeightedModel wModel(1, 0, 1, 0.05, 5);

    // Initialize Kall
    std::vector<std::string> rules = GNM::getRules();
    std::vector<std::vector<std::vector<std::vector<double>>>> Kall(
            1,
            std::vector<std::vector<std::vector<double>>>(
                    rules.size(),
                    std::vector<std::vector<double>>(
                            paramSpace.size(),
                            std::vector<double>(4))));

    // Run the generative models
    int nSamples = 1;
    for (int iSample = 0; iSample < nSamples; ++iSample) {
        for (int jModel = 0; jModel < rules.size(); ++jModel) {
            auto start = std::chrono::high_resolution_clock::now();
            std::vector<std::vector<int>> b(
                    m,
                    std::vector<int>(paramSpace.size())
            );

            std::vector<std::vector<double>> K(
                    paramSpace.size(),
                    std::vector<double>(4)
            );

            WeightedGNM model(
                    A,
                    A_init,
                    D,
                    paramSpace,
                    b,
                    K,
                    m,
                    jModel,
                    paramSpace.size(),
                    A.size(),
                    wModel
            );

            model.generateModels();
            Kall[iSample][jModel] = K;

            auto end = std::chrono::high_resolution_clock::now();
            auto duration = std::chrono::duration_cast<std::chrono::seconds>(end - start);
            std::cout << "Done | Sample: " << iSample << " Model: " << jModel << " Duration: " << duration.count()
                      << std::endl;
        }
    }

    // Store the results
    std::string pOut = "/Users/maxwuerfek/code/diss/gnm-run/testData/testKall.h5";
    GNM::saveResults(pOut, Kall, paramSpace);
    return 0;
}

File: jl/main_test.jl
include("gnm.jl")
include("test_data.jl")
include("gnm_utils.jl")

using .GNM_Mod

function main()
    models = ["spatial", "neighbors", "matching", "clu-avg", "clu-min",
        "clu-max", "clu-dist", "clu-prod", "deg-avg", "deg-min", "deg-max",
        "deg-dist", "deg-prod"
    ]

    As, D, A_inits = load_test_data(2)
    params = generate_param_space(1000)

    K = zeros(Int(size(As, 1)), length(models), Int(size(params, 1)), 4)

    for i_sample in axes(As, 1)
        A = As[i_sample, :, :]
        A_init = A_inits[i_sample, :, :]
        for i_model in 1:length(models)
            println("Sample: ", i_sample, " Model: ", i_model)
            model = GNM_Mod.GNM(A, D, A_init, params, i_model)
            GNM_Mod.generate_models(model)
            K[i_sample, i_model, :, :] = model.K
        end
    end

    out_dir = "/Users/maxwuerfek/code/diss/jl/test/"
    save_K(out_dir, K, params)
end

main()